library/threads.lisp

```diff
@@ -0,0 +1,322 @@
+(coalton-library/utils:defstdlib-package #:coalton-library/threads
+  (:use
+   #:coalton
+   #:coalton-library/system
+   #:coalton-library/classes)
+  (:export
+   ;; Threads
+   #:Thread
+   #:spawn
+   #:make-thread
+   #:current-thread
+   #:all-threads
+   #:join
+   #:interrupt
+   #:destroy
+   #:alive?
+   ;; Locks
+   #:Lock
+   #:RecursiveLock
+   #:with-lock-held
+   #:with-recursive-lock-held
+   #:make-lock
+   #:make-recursive-lock
+   #:acquire-lock
+   #:acquire-lock-no-wait
+   #:acquire-recursive-lock
+   #:acquire-recursive-lock-no-wait
+   #:release-lock
+   #:release-recursive-lock
+   ;; Condition Variables
+   #:ConditionVariable
+   #:make-cv
+   #:await-cv
+   #:notify-cv
+   #:broadcast-cv
+   ;; Semaphores
+   #:Semaphore
+   #:make-semaphore
+   #:signal-semaphore
+   #:await-semaphore
+   ;; Atomics
+   #:AtomicInteger
+   #:make-atomic
+   #:atomic-cmp-and-swap
+   #:decf-atomic
+   #:incf-atomic
+   #:atomic-value))
+
+(in-package #:coalton-library/threads)
+
+(named-readtables:in-readtable coalton:coalton)
+
+#+coalton-release
+(cl:declaim #.coalton-impl/settings:*coalton-optimize-library*)
+
+;;;
+;;; This module implements bindings to bordeaux-threads.
+;;; It will only be loaded if thread capabilities are 
+;;; detected in `cl:*features*'.
+;;;
+
+;;---------;;
+;; Threads ;;
+;;---------;;
+
+(cl:defmacro spawn (cl:&body body)
+  `(make-thread (fn () ,@body)))
+
+(coalton-toplevel
+  (repr :native bt2:thread)
+  (define-type (Thread :a))
+
+  (define-instance (Eq (Thread :a))
+    (define (== a b)
+      (lisp Boolean (a b) (to-boolean (cl:equal a b)))))
+
+  (declare make-thread ((Unit -> :a) -> Thread :a))
+  (define (make-thread thunk)
+    "Creates and returns a thread, which will call the function
+`thunk' with no arguments: when `thunk' returns, the thread terminates."
+    (lisp (Thread :a) (thunk)
+      (bt2:make-thread
+       (cl:lambda () (call-coalton-function thunk Unit)))))
+
+  (declare current-thread (Unit -> Thread :a))
+  (define (current-thread)
```

Comment:
If the current thread is a Thread Integer and the environment unifies :a with Float, won't we have a type error?

I think this is a fundamental issue with this API that can creep in with this function and all below.